### PR TITLE
Disable http/https via environment variables.

### PR DESCRIPTION
**Pull request by iamazeem**
Disable http/https via environment variables.

Signed-off-by: Azeem Sajid <azeem.sajid@gmail.com>

### CODE REVIEW

File: cwm_worker_operator/domains_config.py [hunk 1]
```diff
@@ -144,7 +144,7 @@ def __init__(self, data, domains_config, request_hostname=None, is_data_from_cac
         self.hostname_challenges = {}
         minio_extra_configs = data.get('minio_extra_configs', {})
         self.protocols_enabled = set()
-        for protocol in minio_extra_configs.pop('protocols-enabled', ['http', 'https']):
+        for protocol in minio_extra_configs.pop('protocols-enabled', ['http', 'https', 'disable_http', 'disable_https']):
```
Comment (OriHoch) [inline cwm_worker_operator/domains_config.py:147]: there is no change to the protocols-enabled field, it still contains just http / https (the enabled protocols)

File: cwm_worker_operator/deployer.py [hunk 2]
```diff
@@ -89,6 +89,8 @@ def deploy_worker(domains_config, deployer_metrics, deployments_manager, worker_
             'CDN_CACHE_NOCACHE_REGEX': '\\.({})$'.format('|'.join(volume_config.cache_exclude_extensions)) if len(volume_config.cache_exclude_extensions) > 0 else '',
             'CDN_CACHE_PROXY_CACHE_VALID_200': '{}m'.format(volume_config.cache_expiry_minutes),
             'CDN_CACHE_PROXY_INACTIVE': '{}m'.format(volume_config.cache_expiry_minutes + 1),
+            'DISABLE_HTTP': 'true' if 'disable_http' in volume_config.protocols_enabled else 'false',
+            'DISABLE_HTTPS': 'true' if 'disable_https' in volume_config.protocols_enabled else 'false',
```
Comment (OriHoch) [inline cwm_worker_operator/deployer.py:93]: these fields should be populated like this:
DISABLE_HTTP: 'true' if 'http' not in protocols_enabled
DISABLE_HTTPS: 'true' if 'https' not in protocols_enabled

**Review by OriHoch — CHANGES_REQUESTED** [review]

**Comment by OriHoch** [comment]
please add a test, it should be in test_deployer

File: cwm_worker_operator/deployer.py (hunk 2 above)
Comment (iamazeem) [reply cwm_worker_operator/deployer.py:93]: This won't work. Please see the explanation here: https://stackoverflow.com/a/12199811/7670262.

```python
            'DISABLE_HTTP': 'true' if 'http' not in volume_config.protocols_enabled else 'false',
            'DISABLE_HTTPS': 'true' if 'https' not in volume_config.protocols_enabled else 'false',
```

or,

```
            '{}'.format('http' not in volume_config.protocols_enabled).lower()
            '{}'.format('https' not in volume_config.protocols_enabled).lower()
```

The first one looks more readable.

Comment (OriHoch) [reply cwm_worker_operator/deployer.py:93]: it was just pseudo-code, the important thing is that protocols_enabled should remain as it was

File: cwm_worker_operator/deployer.py [hunk 3]
```diff
@@ -89,6 +89,8 @@ def deploy_worker(domains_config, deployer_metrics, deployments_manager, worker_
             'CDN_CACHE_NOCACHE_REGEX': '\\.({})$'.format('|'.join(volume_config.cache_exclude_extensions)) if len(volume_config.cache_exclude_extensions) > 0 else '',
             'CDN_CACHE_PROXY_CACHE_VALID_200': '{}m'.format(volume_config.cache_expiry_minutes),
             'CDN_CACHE_PROXY_INACTIVE': '{}m'.format(volume_config.cache_expiry_minutes + 1),
+            'DISABLE_HTTP': 'true' if 'http' not in volume_config.protocols_enabled else 'false',
+            'DISABLE_HTTPS': 'true' if 'https' not in volume_config.protocols_enabled else 'false',
```
Comment (OriHoch) [inline cwm_worker_operator/deployer.py:93]: looks like the relevant helm values are boolean and named differently

https://github.com/CloudWebManage/cwm-worker-deployment-minio/blob/main/helm/values.yaml#L356

Comment (iamazeem) [reply cwm_worker_operator/deployer.py:93]: Yes.
That was according to the values mentioned in the issue.
Are you testing that PR?
I'll update the values if you're done with it.
Let me know.

Comment (OriHoch) [reply cwm_worker_operator/deployer.py:93]: I'm not testing it

Comment (iamazeem) [reply cwm_worker_operator/deployer.py:93]: This one? You assigned this to yourself for testing.
https://github.com/CloudWebManage/cwm-worker-cluster/issues/125

Comment (OriHoch) [reply cwm_worker_operator/deployer.py:93]: yes, I will test that one but you should make sure correct values are used

**Comment by iamazeem** [comment]
@OriHoch : CI is failing due to wait timeout. Please see: https://github.com/CloudWebManage/cwm-worker-operator/pull/9/checks?check_run_id=3370105105#step:3:522.

Should I increase the timeout in `bin/prometheus_deploy_wait.sh`?

**Comment by OriHoch** [comment]
> @OriHoch : CI is failing due to wait timeout. Please see: https://github.com/CloudWebManage/cwm-worker-operator/pull/9/checks?check_run_id=3370105105#step:3:522.
> 
> Should I increase the timeout in `bin/prometheus_deploy_wait.sh`?

yes

**Comment by iamazeem** [comment]
@OriHoch : I have been trying to deploy `kube-prometheus-stack` locally:

```shell
helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
helm repo update
helm install prometheus prometheus-community/kube-prometheus-stack --values tests/kube-prometheus-stack.values.yaml
```

The `prometheus-kube-prometheus-operator` pod keeps crashing due to an invalid argument error i.e. `-config-reloader-cpu-request`.

Here are the logs:

```shell
$ kubectl logs prometheus-kube-prometheus-operator-5d5fc68c4c-dlgdl
flag provided but not defined: -config-reloader-cpu-request
Usage of /bin/operator:
  -add_dir_header
    	If true, adds the file directory to the header
  -alertmanager-default-base-image string
    	Alertmanager default base image (path without tag/version) (default "quay.io/prometheus/alertmanager")
  -alertmanager-instance-namespaces value
    	Namespaces where Alertmanager custom resources and corresponding StatefulSets are watched/created. If set this takes precedence over --namespaces or --deny-namespaces for Alertmanager custom resources.
  -alertmanager-instance-selector string
    	Label selector to filter AlertManager Custom Resources to watch.
  -alsologtostderr
    	log to standard error as well as files
  -apiserver string
    	API Server addr, e.g. ' - NOT RECOMMENDED FOR PRODUCTION - http://127.0.0.1:8080'. Omit parameter to run in on-cluster mode and utilize the service account token.
  -ca-file string
    	- NOT RECOMMENDED FOR PRODUCTION - Path to TLS CA file.
  -cert-file string
    	 - NOT RECOMMENDED FOR PRODUCTION - Path to public TLS certificate file.
  -cluster-domain string
    	The domain of the cluster. This is used to generate service FQDNs. If this is not specified, DNS search domain expansion is used instead.
  -config-reloader-cpu string
    	Config Reloader CPU. Value "0" disables it and causes no limit to be configured. (default "100m")
  -config-reloader-image string
    	Reload Image (default "jimmidyson/configmap-reload:v0.4.0")
  -config-reloader-memory string
    	Config Reloader Memory. Value "0" disables it and causes no limit to be configured. (default "25Mi")
  -deny-namespaces value
    	Namespaces not to scope the interaction of the Prometheus Operator (deny list). This is mutually exclusive with --namespaces.
  -key-file string
    	- NOT RECOMMENDED FOR PRODUCTION - Path to private TLS certificate file.
  -kubelet-service string
    	Service/Endpoints object to write kubelets into in format "namespace/name"
  -labels value
    	Labels to be add to all resources created by the operator
  -localhost string
    	EXPERIMENTAL (could be removed in future releases) - Host used to communicate between local services on a pod. Fixes issues where localhost resolves incorrectly. (default "localhost")
  -log-format string
    	Log format to use. Possible values: logfmt, json (default "logfmt")
  -log-level string
    	Log level to use. Possible values: all, debug, info, warn, error, none (default "info")
  -log_backtrace_at value
    	when logging hits line file:N, emit a stack trace
  -log_dir string
    	If non-empty, write log files in this directory
  -log_file string
    	If non-empty, use this log file
  -log_file_max_size uint
    	Defines the maximum size a log file can grow to. Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
  -logtostderr
    	log to standard error instead of files (default true)
  -namespaces value
    	Namespaces to scope the interaction of the Prometheus Operator and the apiserver (allow list). This is mutually exclusive with --deny-namespaces.
  -prometheus-config-reloader string
    	Prometheus config reloader image (default "quay.io/prometheus-operator/prometheus-config-reloader:v0.42.1")
  -prometheus-default-base-image string
    	Prometheus default base image (path without tag/version) (default "quay.io/prometheus/prometheus")
  -prometheus-instance-namespaces value
    	Namespaces where Prometheus custom resources and corresponding Secrets, Configmaps and StatefulSets are watched/created. If set this takes precedence over --namespaces or --deny-namespaces for Prometheus custom resources.
  -prometheus-instance-selector string
    	Label selector to filter Prometheus Custom Resources to watch.
  -secret-field-selector string
    	Field selector to filter Secrets to watch
  -skip_headers
    	If true, avoid header prefixes in the log messages
  -skip_log_headers
    	If true, avoid headers when opening log files
  -stderrthreshold value
    	logs at or above this threshold go to stderr (default 2)
  -thanos-default-base-image string
    	Thanos default base image (path without tag/version) (default "quay.io/thanos/thanos")
  -thanos-ruler-instance-namespaces value
    	Namespaces where ThanosRuler custom resources and corresponding StatefulSets are watched/created. If set this takes precedence over --namespaces or --deny-namespaces for ThanosRuler custom resources.
  -thanos-ruler-instance-selector string
    	Label selector to filter ThanosRuler Custom Resources to watch.
  -tls-insecure
    	- NOT RECOMMENDED FOR PRODUCTION - Don't verify API server's CA certificate.
  -v value
    	number for the log level verbosity
  -vmodule value
    	comma-separated list of pattern=N settings for file-filtered logging
  -web.cert-file string
    	Cert file to be used for operator web server endpoints. (default "/etc/tls/private/tls.crt")
  -web.client-ca-file string
    	Client CA certificate file to be used for operator web server endpoints. (default "/etc/tls/private/tls-ca.crt")
  -web.enable-tls
    	Activate prometheus operator web server TLS.   This is useful for example when using the rule validation webhook.
  -web.key-file string
    	Private key matching the cert file to be used for operator web server endpoints. (default "/etc/tls/private/tls.key")
  -web.listen-address string
    	Address on which to expose metrics and web interface. (default ":8080")
  -web.tls-cipher-suites string
    	Comma-separated list of cipher suites for the server. Values are from tls package constants (https://golang.org/pkg/crypto/tls/#pkg-constants).If omitted, the default Go cipher suites will be used.Note that TLS 1.3 ciphersuites are not configurable.
  -web.tls-min-version string
    	Minimum TLS version supported. Value must match version names from https://golang.org/pkg/crypto/tls/#pkg-constants. (default "VersionTLS13")
  -web.tls-reload-interval duration
    	The interval at which to watch for TLS certificate changes, by default set to 1 minute. (default 1m0s). (default 1m0s)
```

The `prometheus-kube-prometheus-operator` version in [0.42.1](https://github.com/prometheus-operator/prometheus-operator/tree/v0.42.1):

https://github.com/CloudWebManage/cwm-worker-operator/blob/60fa57d44d95a9ac7d6f1e8b5dbcdb91d034f1ee/tests/kube-prometheus-stack.values.yaml#L1362-L1384

And, the flag `-config-reloader-cpu-request` has not been defined in [0.42.1](https://github.com/prometheus-operator/prometheus-operator/tree/v0.42.1):

See: https://github.com/prometheus-operator/prometheus-operator/blob/a1eb8fd04f3cd2bfb11aee8943e0b3469ff2f199/cmd/operator/main.go#L169-L189

Do you also observe this locally with minikube?

**Comment by iamazeem** [comment]
Tested with the latest [0.50.0](https://github.com/prometheus-operator/prometheus-operator/tree/v0.50.0) version and it deployed successfully.

```text
NAMESPACE              NAME                                                   READY   STATUS    RESTARTS   AGE
default                prometheus-kube-prometheus-operator-7988d98b77-dq69c   1/1     Running   0          4m47s
default                prometheus-kube-state-metrics-c69d6cddc-rbcqm          1/1     Running   0          4m47s
default                prometheus-prometheus-kube-prometheus-prometheus-0     2/2     Running   0          3m9s
default                prometheus-prometheus-node-exporter-vw7jv              1/1     Running   0          4m46s
kube-system            calico-kube-controllers-77d6cbc65f-2c48s               1/1     Running   17         86d
kube-system            calico-node-t7v9j                                      1/1     Running   20         86d
kube-system            coredns-66bff467f8-vc8fz                               1/1     Running   10         29d
kube-system            etcd-minikube                                          1/1     Running   13         29d
kube-system            kube-apiserver-minikube                                1/1     Running   13         29d
kube-system            kube-controller-manager-minikube                       1/1     Running   14         29d
kube-system            kube-ingress-dns-minikube                              1/1     Running   6          7d21h
kube-system            kube-proxy-k9kcr                                       1/1     Running   12         29d
kube-system            kube-scheduler-minikube                                1/1     Running   13         29d
kube-system            storage-provisioner                                    1/1     Running   111        190d
kubernetes-dashboard   dashboard-metrics-scraper-b68468655-pqv6r              1/1     Running   18         86d
kubernetes-dashboard   kubernetes-dashboard-7dcf795d9f-hg7kx                  1/1     Running   18         86d
```

**Review by OriHoch — APPROVED** [review]

**Pull request merged**